test/SILOptimizer/opt-remark-generator.swift

```diff
@@ -70,3 +20,223 @@ public func useGlobal() {
+             // expected-note @-4:26 {{of 'x'}}
+             // expected-remark @-2:5 {{retain}}
+             // expected-note @-6:26 {{of 'x'}}
+             // Release on temp array for print(...).
+             // expected-remark @-5:12 {{release}}
+}
+
+func printKlassTupleLHS(x : (Klass, Klass)) {
+    // We print the remarks at the 'p' and at the ending ')'.
+    print(x.0) // expected-remark @:5 {{retain}}
+               // expected-note @-3:25 {{of 'x'}}
+               // Release on Array<Any> for print.
+               // expected-remark @-3:14 {{release}}
+}
+
+func returnKlassTupleLHS(x: (Klass, Klass)) -> Klass {
+    return x.0 // expected-remark @:12 {{retain}}
+               // expected-note @-2:26 {{of 'x'}}
+}
+
+func callingAnInitializerKlassTuplePair(x: Klass, y: Klass) -> (Klass, Klass) {
+    return (x, y) // expected-remark {{retain}}
+                  // expected-note @-2:41 {{of 'x'}}
+                  // expected-remark @-2:5 {{retain}}
+                  // expected-note @-4:51 {{of 'y'}}
+}
+
+public class SubKlass : Klass {
+    @inline(never)
+    final func doSomething() {}
+}
+
+func lookThroughCast(x: SubKlass) -> Klass {
+    return x as Klass // expected-remark {{retain}}
+                      // expected-note @-2:22 {{of 'x'}}
+}
+
+func lookThroughRefCast(x: Klass) -> SubKlass {
+    return x as! SubKlass // expected-remark {{retain}}
+                          // expected-note @-2:25 {{of 'x'}}
+}
+
+func lookThroughEnum(x: Klass?) -> Klass {
+    return x! // expected-remark {{retain}}
+              // expected-note @-2:22 {{of 'x.some'}}
+}
+
+func castAsQuestion(x: Klass) -> SubKlass? {
+    x as? SubKlass // expected-remark {{retain}}
+                   // expected-note @-2:21 {{of 'x'}}
+}
+
+func castAsQuestionDiamond(x: Klass) -> SubKlass? {
+    guard let y = x as? SubKlass else {
+        return nil
+    }
+
+    y.doSomething()
+    return y // expected-remark {{retain}}
+             // expected-note @-7:28 {{of 'x'}}
+}
+
+func castAsQuestionDiamondGEP(x: KlassPair) -> SubKlass? {
+    guard let y = x.lhs as? SubKlass else {
+        return nil
+    }
+
+    y.doSomething()
+    // We eliminate the rhs retain/release.
+    return y // expected-remark {{retain}}
+             // expected-note @-8:31 {{of 'x.lhs'}}
+}
+
+// We don't handle this test case as well.
+func castAsQuestionDiamondGEP2(x: KlassPair) {
+    switch (x.lhs as? SubKlass, x.rhs as? SubKlass) { // expected-remark @:19 {{retain}}
+                                                      // expected-note @-2 {{of 'x.lhs'}}
+                                                      // expected-remark @-2:39 {{retain}}
+                                                      // expected-note @-4 {{of 'x.rhs'}}
+    case let (.some(x1), .some(x2)):
+        print(x1, x2) // expected-remark {{retain}}
+                      // expected-remark @-1 {{retain}}
+                      // expected-remark @-2 {{release}}
```

Comment:
I am going to fix this in a follow on commit. I just need to recognize enum phi arguments.